File: src/main/java/cf/jrozen/mh/ttp/util/Loader.java
package cf.jrozen.mh.ttp.util;

import cf.jrozen.mh.ttp.model.Node;
import cf.jrozen.mh.ttp.model.Problem;
import cf.jrozen.mh.ttp.model.Section;

import java.util.Arrays;
import java.util.List;
import java.util.Scanner;
import java.util.stream.Collectors;

public class Loader {

    private static final String DIR_NAME = "fixtures/";
    private static final String FILE_EXT = ".ttp";
    public static final String REGEX = ":\\p{javaSpaceChar}{2,}";

    public static Problem load(String filename) {
        final Scanner scanner = new Scanner(Loader.class.getClassLoader().getResourceAsStream(DIR_NAME + filename + FILE_EXT));

        final String problemName = scanner.nextLine().replaceAll(REGEX, ":").split(":")[1];
        final String knapsackDataType = scanner.nextLine().replaceAll(REGEX, ":").split(":")[1];
        final int dimension = Integer.parseInt(scanner.nextLine().replaceAll(REGEX, ":").split(":")[1]);
        final int numberOfItems = Integer.parseInt(scanner.nextLine().replaceAll(REGEX, ":").split(":")[1]);
        final int capacityOfKnapsack = Integer.parseInt(scanner.nextLine().replaceAll(REGEX, ":").split(":")[1]);
        final double minSpeed = Double.parseDouble(scanner.nextLine().replaceAll(REGEX, ":").split(":")[1]);
        final double maxSpeed = Double.parseDouble(scanner.nextLine().replaceAll(REGEX, ":").split(":")[1]);
        final double rentingRatio = Double.parseDouble(scanner.nextLine().replaceAll(REGEX, ":").split(":")[1]);
        final String edgeWeightType = scanner.nextLine().replaceAll(REGEX, ":").split(":")[1];

        final List<Node> nodes = scanner.tokens().takeWhile("ITEMS SECTION"::contains).map(Loader::nodeFromLine).collect(Collectors.toList());
        final List<Section> section = scanner.tokens().map(Loader::sectionFromLine).collect(Collectors.toList());

        return new Problem(
                problemName, knapsackDataType, dimension, numberOfItems, capacityOfKnapsack,
                minSpeed, maxSpeed, rentingRatio, edgeWeightType, nodes, section
        );
    }

    private static Node nodeFromLine(String line) {
        final double[] ints = Arrays.stream(line.split(REGEX)).mapToDouble(Double::parseDouble).toArray();
        return new Node((int) ints[0], ints[1], ints[2]);
    }

    private static Section sectionFromLine(String line) {
        final double[] ints = Arrays.stream(line.split(REGEX)).mapToDouble(Double::parseDouble).toArray();
        return new Section((int) ints[0], ints[1], ints[2], (int) ints[3]);
    }
}

File: src/main/scala/cf/jrozen/mh/ttp/utils/Loader.scala
package cf.jrozen.mh.ttp.utils

import cf.jrozen.mh.ttp.model.{Item, Node, Problem}

import scala.io.Source

object Loader {

  private val DIR_NAME = "fixtures/"
  private val FILE_EXT = ".ttp"
  private val REGEX = ":"

  def load(filename: String): Problem = {

    val scanner: Iterator[String] = Source.fromInputStream(inputStream(filename)).getLines()

    val problemName = nextValue(scanner)
    val knapsackDataType = nextValue(scanner)
    val dimension = nextValue(scanner).toInt
    val numberOfItems = nextValue(scanner).toInt
    val capacityOfKnapsack = nextValue(scanner).toInt
    val minSpeed = nextValue(scanner).toDouble
    val maxSpeed = nextValue(scanner).toDouble
    val rentingRatio = nextValue(scanner).toDouble
    val edgeWeightType = nextValue(scanner)

    scanner.next()

    val nodes = scanner.takeWhile(notContainsItemsSection).map(nodeFromLine).toArray
    val sections = scanner.map(itemFromLine).toArray

    Problem(problemName, knapsackDataType, dimension, numberOfItems, capacityOfKnapsack, minSpeed, maxSpeed, rentingRatio, edgeWeightType, nodes, sections)
  }

  val notContainsItemsSection: String => Boolean = str => !(str contains "ITEMS SECTION")

  private def inputStream(filename: String) =
    classOf[Problem].getClassLoader.getResourceAsStream(DIR_NAME + filename + FILE_EXT)


  private def nextValue(scanner: Iterator[String]) = scanner.next().split(REGEX)(1).trim

  private def nodeFromLine(line: String) = {
    val values = line.split("\t").map(_.trim.toDouble)
    Node(values(0).toInt, values(1), values(2))
  }

  private def itemFromLine(line: String) = {
    val values = line.split("\t").map(_.trim.toDouble)
    Item(values(0).toInt, values(1), values(2), values(3).toInt)
  }

}
